Cross compilers

giggsoff opened the pull request:
Let's enable cross compilation for kernel and new-kernel (from amd64 host for now). It reduce kernel compilation time arm64 target on my PC (amd64) from more than 2 hours down to 15 minutes.

The PR contains new packages where we build apks for cross-compilers using slightly modified aports with the same versions we use for native compilers. The build process of new package is quite time consuming, but we can just remove it from our build system as well as we do it for alpine-base until we will move to another version of Alpine.

Inside kernel and new-kernel we install cross-compilers and extract libraries to sysroot from native eve-alpine.
Only one change required in kernel config to use cross-compilers (CONFIG_CC_VERSION_TEXT from just gcc in the beginning to arch-specific).

deitch commented:
I am trying to wrap my head around this, so forgive the longer questions.

Not all that many files are changed. The `Makefile` and `pkg/alpine/mirrors/3.16/main` are straightforward. That leaves just the new `pkg/cross-compilers` and changes to the `pkg/{kernel,new-kernel}/Dockerfile`.

### cross-compilers

There are 3 `FROM`, one each for `amd64`, `arm64`, `riscv64`. The latter two have, well, nothing in them that isn't in `lfedge/eve-alpine` already, while the `amd64` one has lots on it. It looks like it builds _everything_ in `aports` after cloning? But first applies a few patches?

I don't understand:

* what the purpose of `cross-compilers` is (don't the native gcc and other tools already support cross-compiling)?
* if this will be a very long and expensive operation, which will need to pull tons of things from the Internet?
* why there is this work in amd64 only?

### kernels

I don't understand what the flow is here. The current one is fairly straightforward. Most of the `FROM` images are really just setting some vars, and then it all gets used in a single `kernel-build`

```

                                    |---> kernel-base-amd64. |
kernel-build-base ------> kernel-base-arm64    ----> kernel-build
                                    |---> kernel-base-riscv64  |
```

This appears to have a whole series of "build from A for B", "build from B for A", etc. It almost looks like it has a fanout (that is too hard for me to diagram above).

Is that necessary? Can we not just say, "if target != build then ...."?

### running in different locations

what happens when I build on arm64? Does it try to build for arm64? For amd64? For all, none? What about pushing?

giggsoff commented:
> I don't understand:
> 
> * what the purpose of `cross-compilers` is (don't the native gcc and other tools already support cross-compiling)?

We should prepare specific compiler for out target. It is not native one.

> * if this will be a very long and expensive operation, which will need to pull tons of things from the Internet?

The main idea of the patches I put is to adjust the [bootstrap](https://gitlab.alpinelinux.org/alpine/aports/-/blob/master/scripts/bootstrap.sh) process which exists in Alpine. Bootstrap is the script to build packages for another arch. I just want to stop the script right after building of cross-compilers and before going into packages build. This way we will have cross-compilers to be based on the same version and patches as our native compilers. The time is not so dramatic (on my PC it is twice faster than build of kernel for arm64 in emulation mode).

> * why there is this work in amd64 only?

Actually I want to speedup my own builds and builds in CI (riscv64 now uses emulation). I hit errors when building for riscv64 host. For arm64 host I can try, but I cannot check how it will work on my PC without another layer of emulation...

> ### kernels
> I don't understand what the flow is here. The current one is fairly straightforward. Most of the `FROM` images are really just setting some vars, and then it all gets used in a single `kernel-build`
> 
> ```
> 
>                                     |---> kernel-base-amd64. |
> kernel-build-base ------> kernel-base-arm64    ----> kernel-build
>                                     |---> kernel-base-riscv64  |
> ```
> 
> This appears to have a whole series of "build from A for B", "build from B for A", etc. It almost looks like it has a fanout (that is too hard for me to diagram above).
> 
> Is that necessary? Can we not just say, "if target != build then ...."?

I forced to use different base images to fit host and pull libraries for target arch in place. It might looks better in case of support for all cross-compilers, but at least riscv64 have problems.

> ### running in different locations
> what happens when I build on arm64? Does it try to build for arm64? For amd64? For all, none? What about pushing?

Let me describe what changed:
when you are on amd64 host, and want to build kernel for arm64 or riscv64 you will use cross-compilation (you can see this in build workflow for riscv64).

uncleDecart commented:
I think @rene might have some interesting points to add to this discussion regarding arm cross-compilation

deitch commented:
> We should prepare specific compiler for out target. It is not native one.

The default `gcc` and `binutils` package for alpine are only for the local arch; this is to set up the gcc cross-compiler? OK.

> The main idea of the patches I put is to adjust the [bootstrap](https://gitlab.alpinelinux.org/alpine/aports/-/blob/master/scripts/bootstrap.sh) process which exists in Alpine. Bootstrap is the script to build packages for another arch. I just want to stop the script right after building of cross-compilers and before going into packages build.

Now that you described it, I look in and see it. That leaves a few questions:

1. It does not build all of the massive aports packages, about 68 of them. Are those not necessary for cross-compilation?
2. If we only need that part, rather than a custom patch to maintain it, can we upstream an option to exit after the cross-compiler creation? We don't want yet another patch to maintain.
3. What is the second patch with sysroot?

> Actually I want to speedup my own builds and builds in CI (riscv64 now uses emulation). I hit errors when building for riscv64 host. For arm64 host I can try, but I cannot check how it will work on my PC without another layer of emulation

The assumption then becomes, "the default is amd64", which is not great. Can we have it working on all supported platforms? Sure, we cannot test riscv64, but we don't have any riscv64 platforms anyways. We should try to get it as uniform as possible.

Actually, looking at `pkg/cross-compilers/Dockerfile`, there is almost nothing "amd64-unique" in that section, just the targets. Can we have the same code for all platforms, with a switch or if statement that determines which platforms to target (everything except myself)? Something sort of like:

```dockerfile
ENV allTargets="riscv64 aarch64 x86_64"
RUN busybox su builder -c 'for patch in /home/builder/patches/*patch ; do patch -p1 < $patch ; done' && \
    mine=$(name -m); && \
    for target in ${allTargets}; do \
      [ "${target}" = "${mine}" ] && continue; && \
       busybox su builder -c 'sh -x ./scripts/bootstrap.sh ${target} && \
    done; && \
     rm -rf /home/builder/packages/main/"$(uname -m)"/gcc-pass2* && \
     mkdir /packages && \
     cp -r /home/builder/packages/main/"$(uname -m)" /packages/
```

> I forced to use different base images to fit host and pull libraries for target arch in place. It might looks better in case of support for all cross-compilers, but at least riscv64 have problems.


If you are saying it is necessary, then so be it. Can we describe the flow in documents? When it gets that convoluted, it will be hard for the next person to think about.

I think the stages are the following. For the examples, I am on B building for T

* `kernel-build-base-native` - T
* `kernel-build-base-cross` - B
* `cross-compilers`  - just to get my cross compilers, B
* `cross-compile-libs` - just to get my linking libraries, which must be native to the target, so T
* `kernel-cross-target-arm64`, `kernel-cross-target-amd64`, `kernel-cross-target-riscv64` - just to set the `TARGET_ARCH` env var - do we need this? Is this because the built-in `TARGETARCH` uses `arm64` instead of `aarch64`, etc.? I also do not see `kernel-cross-target-amd64`
* `kernel-cross-build-amd64`, `kernel-cross-build-arm64`, `kernel-cross-build-riscv64` - First, I don't see the `arm64` and `riscv64` variants of this. Second, is this just for installing the necessary binaries and some more env vars?
* `kernel-target-arm64-build-amd64` (and the 8 variants on combining target and build, even if not all of them are there) - I lost track of what this one does that previous stages do not
* `kernel-target-amd64`, `kernel-target-amd64`, `kernel-target-riscv64` - some more vars? Not really sure.
* `kernel-build` - this is the one that does the final build

That is a lot of interim stages. Which isn't a problem by itself (although it always ie better if simple), as long as it is understood and well documented.

giggsoff commented:
> Now that you described it, I look in and see it. That leaves a few questions:
> 
> 1. It does not build all of the massive aports packages, about 68 of them. Are those not necessary for cross-compilation?

No, other packages are available in eve-alpine for target arch.

> 2. If we only need that part, rather than a custom patch to maintain it, can we upstream an option to exit after the cross-compiler creation? We don't want yet another patch to maintain.

We want to use the same tag (3.16.2) as we use for eve-alpine image. I do not thing that it is possible to have any patches there.

> 3. What is the second patch with sysroot?

Without the patch sysroot of resulting binary will be pointed onto non-existing directory ($CBUILDROOT) by default. That directory exists only inside of alpine bootstrap process. And resulting cross-compiler use that directory to find required libraries/headers during bootstrap. But we want to have another sysroot.

> Actually, looking at `pkg/cross-compilers/Dockerfile`, there is almost nothing "amd64-unique" in that section, just the targets. Can we have the same code for all platforms, with a switch or if statement that determines which platforms to target (everything except myself)? Something sort of like:
> 
> ```dockerfile
> ENV allTargets="riscv64 aarch64 x86_64"
> RUN busybox su builder -c 'for patch in /home/builder/patches/*patch ; do patch -p1 < $patch ; done' && \
>     mine=$(name -m); && \
>     for target in ${allTargets}; do \
>       [ "${target}" = "${mine}" ] && continue; && \
>        busybox su builder -c 'sh -x ./scripts/bootstrap.sh ${target} && \
>     done; && \
>      rm -rf /home/builder/packages/main/"$(uname -m)"/gcc-pass2* && \
>      mkdir /packages && \
>      cp -r /home/builder/packages/main/"$(uname -m)" /packages/
> ```

Will modify, thank you.

> I think the stages are the following. For the examples, I am on B building for T
> 
> * `kernel-build-base-native` - T
> * `kernel-build-base-cross` - B
> * `cross-compilers`  - just to get my cross compilers, B
> * `cross-compile-libs` - just to get my linking libraries, which must be native to the target, so T
> * `kernel-cross-target-arm64`, `kernel-cross-target-amd64`, `kernel-cross-target-riscv64` - just to set the `TARGET_ARCH` env var - do we need this? Is this because the built-in `TARGETARCH` uses `arm64` instead of `aarch64`, etc.? I also do not see `kernel-cross-target-amd64`

Correct, the problem is in arm64/aarch64. You cannot see `kernel-cross-target-amd64` because we do not have any cross-compilers for amd64 target (only riscv64 and arm64 targets for build=amd64).

> * `kernel-cross-build-amd64`, `kernel-cross-build-arm64`, `kernel-cross-build-riscv64` - First, I don't see the `arm64` and `riscv64` variants of this. Second, is this just for installing the necessary binaries and some more env vars?

You cannot see `arm64` and `riscv64` here because we do not have any cross-compilers with build=arm64 or build=riscv64, only build=amd64.

> * `kernel-target-arm64-build-amd64` (and the 8 variants on combining target and build, even if not all of them are there) - I lost track of what this one does that previous stages do not

Just ENV adjustment. Probably we can move such ENVs into pkg/alpine and pkg/cross-compilers

> * `kernel-target-amd64`, `kernel-target-amd64`, `kernel-target-riscv64` - some more vars? Not really sure.

Correct, because kernel use x86 instead of amd64 and x86_64...

> That is a lot of interim stages. Which isn't a problem by itself (although it always ie better if simple), as long as it is understood and well documented.

Do you have any good place to put the document in mind?

deitch commented:
> No, other packages are available in eve-alpine for target arch.

Understood. So the `bootstrap.sh` really is meant as "bootstrap everything as if nothing existed", and we are working in a case where many already exist.

What is the percentage difference in execution time between just running `bootstrap.sh` as is vs adding the exit patch? It is inefficient to build them all, but then again, so is maintaining a custom patch.

> We want to use the same tag (3.16.2) as we use for eve-alpine image. I do not thing that it is possible to have any patches there.

Do you mean that we _could_ submit a patch, but it only would go in for future releases? Are we going to try? Sometimes they do back port to existing releases.

> Without the patch sysroot of resulting binary will be pointed onto non-existing directory ($CBUILDROOT) by default. That directory exists only inside of alpine bootstrap process. And resulting cross-compiler use that directory to find required libraries/headers during bootstrap. But we want to have another sysroot.

But isn't that generated compiler meant to be used to build all packages? How do all of the other packages use it?

Yes, I am working quite hard to avoid patches.

> You cannot see kernel-cross-target-amd64 because we do not have any cross-compilers for amd64 target (only riscv64 and arm64 targets for build=amd64

We might as well make it all arches. Tomorrow (today?) someone will want to run locally on a Mac M1 to build for x86_64, etc. More importantly, people will be surprised in 3 months when they try to run it and it doesn't work, and have to dig deeply into this to understand why.

> Correct, the problem is in arm64/aarch64

What a mess. I wonder why the buildkit/docker folks didn't make `TARGETARCH` and maybe `TARGETUNAMEM` or something available, so you could have it both ways.

Maybe this is another one you could eliminate. Instead of:

```dockerfile
FROM kernel-build-base-cross AS kernel-cross-target-arm64
ENV TARGET_ARCH=aarch64

# install cross-compile packages and libs for target
# hadolint ignore=DL3006
FROM kernel-cross-target-${TARGETARCH} AS kernel-cross-build-amd64
ENV BUILD_ARCH=x86_64
ENV CROSS_COMPILE_ENV="${TARGET_ARCH}"-alpine-linux-musl-
COPY --from=cross-compilers /packages /packages
# hadolint ignore=DL3018
RUN apk add --no-cache --allow-untrusted -X /packages build-base-"${TARGET_ARCH}"
COPY --from=cross-compile-libs /out/ /usr/"${TARGET_ARCH}"-alpine-linux-musl/

# support cross compile from amd64 for now
# hadolint ignore=DL3006
FROM kernel-cross-build-${BUILDARCH} AS kernel-target-arm64-build-amd64

# set host and target to target for the reset of stages (native om emulated)
FROM kernel-build-base-native AS kernel-target-amd64-build-amd64
ENV BUILD_ARCH=x86_64
ENV TARGET_ARCH=x86_64
FROM kernel-build-base-native AS kernel-target-amd64-build-arm64
ENV BUILD_ARCH=x86_64
ENV TARGET_ARCH=x86_64
FROM kernel-build-base-native AS kernel-target-arm64-build-arm64
ENV BUILD_ARCH=aarch64
ENV TARGET_ARCH=aarch64
```

Maybe you can simplify all of those as:

```dockerfile
# set host and target for all stages
FROM kernel-build-base-native AS kernel-target-amd64
ENV TARGET_ARCH=x86_64
# additional ENV settings for amd64 targets

FROM kernel-build-base-native AS kernel-target-arm64
ENV TARGET_ARCH=aarch64
# additional ENV settings for arm64 targets

FROM kernel-build-base-native AS kernel-target-riscv64
ENV TARGET_ARCH=riscv64
# additional ENV settings for riscv64 targets

FROM kernel-target-${TARGET_ARCH) AS kernel-build-amd64
ENV BUILD_ARCH=x86_64
FROM kernel-target-${TARGET_ARCH) AS kernel-build-arm64
ENV BUILD_ARCH=aarch64
FROM kernel-target-${TARGET_ARCH) AS kernel-build-riscv64
ENV BUILD_ARCH=riscv64

# hadolint ignore=DL3006
FROM kernel-build-${BUILDARCH} AS kernel-cross-build
ENV CROSS_COMPILE_ENV="${TARGET_ARCH}"-alpine-linux-musl-
COPY --from=cross-compilers /packages /packages
# hadolint ignore=DL3018
RUN apk add --no-cache --allow-untrusted -X /packages build-base-"${TARGET_ARCH}"
COPY --from=cross-compile-libs /out/ /usr/"${TARGET_ARCH}"-alpine-linux-musl/

# etc etc
```

It just reduces an extra stage there.

> AS cross-compile-libs

Why bother? This just is the same source as `kernel-build-base-native`. We can simplify the current:

```Dockerfile
FROM lfedge/eve-alpine:1fc941603ddc153f26104865db242e644c16aa99 as kernel-build-base-native
ARG BUILD_PKGS_BASE
RUN BUILD_PKGS="${BUILD_PKGS_BASE}" eve-alpine-deploy.sh

# ...

FROM lfedge/eve-alpine:1fc941603ddc153f26104865db242e644c16aa99 AS cross-compile-libs
ENV PKGS musl-dev libgcc
RUN eve-alpine-deploy.sh

# ...

FROM kernel-cross-target-${TARGETARCH} AS kernel-cross-build-amd64
# ...
COPY --from=cross-compile-libs /out/ /usr/"${TARGET_ARCH}"-alpine-linux-musl/
```

to

```Dockerfile
FROM lfedge/eve-alpine:1fc941603ddc153f26104865db242e644c16aa99 as kernel-build-base-native
ARG BUILD_PKGS_BASE
ENV PKGS musl-dev libgcc 
RUN BUILD_PKGS="${BUILD_PKGS_BASE}" eve-alpine-deploy.sh

# ...

FROM kernel-cross-target-${TARGETARCH} AS kernel-cross-build-amd64
# ...
COPY --from=kernel-build-base-native /out/usr/"${TARGET_ARCH}"-alpine-linux-musl/ /usr/"${TARGET_ARCH}"-alpine-linux-musl/
```

I got that a bit wrong, but the idea is the same. Let's get rid of one unnecessary interim stage that uses the identical source as the previous.

deitch commented:
> Do you have any good place to put the document in mind?

I think this is worthy of its own document in `docs/CROSS-COMPILE.md` , which, for now, only has kernel in it.

deitch commented:
To be clear, I really like what you are doing. The more cross-compilation we do, the better and faster eve gets.

giggsoff commented:
> But isn't that generated compiler meant to be used to build all packages? How do all of the other packages use it?

All logic there use the directory to extract dependencies into. The directory is something like `/home/builder/sysroot-riscv64` which is quite useless from external point of view.

> We might as well make it all arches. Tomorrow (today?) someone will want to run locally on a Mac M1 to build for x86_64, etc. More importantly, people will be surprised in 3 months when they try to run it and it doesn't work, and have to dig deeply into this to understand why.

I made the things to work in all combinations. So we will not surprise somebody sitting with Mac M1, he will use emulation as we do it now.

As I mentioned previously, building for riscv64 host will require additional patches (disabling of Ada language is a simple one, another one is not clear to me right now, but as I cannot build with disabling of Ada in my experiments, we need another patch). Do you want me to prepare patches?

> It just reduces an extra stage there.

I use different base image `FROM kernel-build-base-cross AS kernel-cross-target-arm64` pointed onto BUILDPLATFORM, so it is not possible to simplify this way.

> Why bother? This just is the same source as kernel-build-base-native. We can simplify the current:

We do not want to add something into /out as we copy from there in the last stage. Moreover, I want to have control and visibility over our dependencies in cross-compile-libs, so it looks like complication of logic to me.

giggsoff commented:
I moved declaration of BUILD_ARCH and TARGET_ARCH into eve-alpine.
Also modified the logic to build arm64 based cross-compiler as well, let's see how it works in our builders.

rene commented on pkg/kernel/Dockerfile:
```diff
@@ -1,33 +1,79 @@
+
+# install cross-compile packages and libs for target
+# hadolint ignore=DL3006
+FROM kernel-cross-target-${TARGETARCH} AS kernel-cross-build-amd64
```
don't you need to use --platform=${BUILDPLATFORM} here as well? Are you installing cross compiler + libs into one single container, right?

giggsoff replied:
Correct, I install both into the image which based on `kernel-cross-target-${TARGETARCH}`, which one in turns based on `kernel-build-base-cross` and the last one based on `FROM --platform=${BUILDPLATFORM} lfedge/eve-alpine:3e225b14d659519459aefbd77cf9eb25070df594`. So, all of them inherit `--platform=${BUILDPLATFORM}`.

rene commented on pkg/kernel/Dockerfile:
```diff
@@ -1,33 +1,79 @@
+FROM kernel-cross-build-${BUILDARCH} AS kernel-target-arm64-build-amd64
+
+# set host and target to target for the reset of stages (native om emulated)
+FROM kernel-build-base-native AS kernel-target-amd64-build-amd64
```
You have all these variables already available from docker:

```
BUILDPLATFORM — matches the current machine. (e.g. linux/amd64)

BUILDOS — os component of BUILDPLATFORM, e.g. linux

BUILDARCH — e.g. amd64, arm64, riscv64

BUILDVARIANT — used to set ARM variant, e.g. v7

TARGETPLATFORM — The value set with --platform flag on build

TARGETOS - OS component from --platform, e.g. linux

TARGETARCH - Architecture from --platform, e.g. arm64
```
It isn't possible to simplify these stuff here?

giggsoff replied:
The problem is about arm64->aarch64 conversion. But I tried to simplify the logic by moving `BUILD_ARCH` and `TARGET_ARCH` into eve-alpine (thanks @Insei for the idea) inside https://github.com/lf-edge/eve/pull/2978/commits/0c0743aad70354db8d15a5ed2a74b503df70a6bc. So now we set only few ENVs and most of the logic is about chaining from cross or native base image.

deitch replied:
> The problem is about arm64->aarch64 conversion. But I tried to simplify the logic by moving BUILD_ARCH and TARGET_ARCH into eve-alpine (thanks @Insei for the idea) inside https://github.com/lf-edge/eve/commit/0c0743aad70354db8d15a5ed2a74b503df70a6bc.

That is a _really_ nice idea. Simple and elegant, and everything gets it.

I will ping the buildkit folks about this too.

rene commented:
@giggsoff , Indeed cross-compilation it's a must and I've been looking around specially because of the current painful process of building the kernel for ARM from x86. I think @deitch pointed out many questions that I was also wondering myself. Nevertheless I would like to make some considerations as well....

- By using FROM --platform=.... you are enforcing the container to enter in the specified platform even if we are building for another platform, so you can do the build process in host platform (cross compiling) and switch to emulation layer in order to do the chroot and run necessary commands (installation, etc). This approach can be followed by many packages as possible, so it could significantly speed up the whole build process, not only kernel. But of course, start by the kernel package it's a good start. Please, keep that in mind. For example, what is the effort to use cross-compiler package to build also go packages? Do many things needs to be added?

- I would split the cross-compilers package, for example, I don't care about riscv64, but the compiler it's going to be built in my host anyways, right? (https://github.com/lf-edge/eve/pull/2978/commits/6008f9e60a2dc88a1c4eedab7c32393cf7767d3c#diff-c25175bfc882ae7db60a6148c29c436dbbf12d10a9bd43935451e5c9969ad07b) Is that possible to build the cross-compiler only to the target platform?

- ARM64 hosts must be considered (so it will be needed to build x86 cross compiler), if you can't test at your side, let's sync up, I can find you people with Macs M1's and M2's to test your changes...

- Please, add information to your commit messages, your commits are: 1) introducing new packages; 2) introducing cross-compilers; 3) changing the way kernel is built.... so please, give as much information as possible regarding your changes so one doesn't need to go through the sources to understand the overview of the changes. Also, information about cross-compilation must be added to the documentation (I would recommend to add them to https://github.com/lf-edge/eve/blob/master/docs/BUILD.md)

giggsoff commented:
Thank you for comments and sorry for the delayed response.

> * By using FROM --platform=.... you are enforcing the container to enter in the specified platform even if we are building for another platform, so you can do the build process in host platform (cross compiling) and switch to emulation layer in order to do the chroot and run necessary commands (installation, etc). This approach can be followed by many packages as possible, so it could significantly speed up the whole build process, not only kernel. But of course, start by the kernel package it's a good start. Please, keep that in mind. For example, what is the effort to use cross-compiler package to build also go packages? Do many things needs to be added?

I modified the pillar to use cross-compilation. Seems, the logic in cross-compilers itself works fine. Actually, But I do not want to modify every package, only packages that consume enormous time, probably xen-tools is a good next candidate.

> * I would split the cross-compilers package, for example, I don't care about riscv64, but the compiler it's going to be built in my host anyways, right? ([6008f9e#diff-c25175bfc882ae7db60a6148c29c436dbbf12d10a9bd43935451e5c9969ad07b](https://github.com/lf-edge/eve/commit/6008f9e60a2dc88a1c4eedab7c32393cf7767d3c#diff-c25175bfc882ae7db60a6148c29c436dbbf12d10a9bd43935451e5c9969ad07b)) Is that possible to build the cross-compiler only to the target platform?

Not sure, that it is possible without inventing of other packages with arch in name.
My idea is to exclude cross-compilers from CI and use them pre-builded. Actually their update (once stabilized for sure) required only in case of moving into another release of Alpine.

> * ARM64 hosts must be considered (so it will be needed to build x86 cross compiler), if you can't test at your side, let's sync up, I can find you people with Macs M1's and M2's to test your changes...

I finished experiments on Arm64 platform, resolved the bug and now it should work fine. But if you will find anybody with M1, it would be great to check. Something like `make pkg/alpine pkg/cross-compilers&&make ZARCH=amd64 pkg/alpine pkg/kernel pkg/new-kernel pkg/pillar` should use cross-compilation

> * Please, add information to your commit messages, your commits are: 1) introducing new packages; 2) introducing cross-compilers; 3) changing the way kernel is built.... so please, give as much information as possible regarding your changes so one doesn't need to go through the sources to understand the overview of the changes. Also, information about cross-compilation must be added to the documentation (I would recommend to add them to https://github.com/lf-edge/eve/blob/master/docs/BUILD.md)

Sure, will do.

rene commented:
> Thank you for comments and sorry for the delayed response.
> 
> > * By using FROM --platform=.... you are enforcing the container to enter in the specified platform even if we are building for another platform, so you can do the build process in host platform (cross compiling) and switch to emulation layer in order to do the chroot and run necessary commands (installation, etc). This approach can be followed by many packages as possible, so it could significantly speed up the whole build process, not only kernel. But of course, start by the kernel package it's a good start. Please, keep that in mind. For example, what is the effort to use cross-compiler package to build also go packages? Do many things needs to be added?
> 
> I modified the pillar to use cross-compilation. Seems, the logic in cross-compilers itself works fine. Actually, I do not want to modify every package, only packages that consume enormous time, probably xen-tools is a good next candidate.
> 
> > * I would split the cross-compilers package, for example, I don't care about riscv64, but the compiler it's going to be built in my host anyways, right? ([6008f9e#diff-c25175bfc882ae7db60a6148c29c436dbbf12d10a9bd43935451e5c9969ad07b](https://github.com/lf-edge/eve/commit/6008f9e60a2dc88a1c4eedab7c32393cf7767d3c#diff-c25175bfc882ae7db60a6148c29c436dbbf12d10a9bd43935451e5c9969ad07b)) Is that possible to build the cross-compiler only to the target platform?
> 
> Not sure, that it is possible without inventing of other packages with arch in name. My idea is to exclude cross-compilers from CI and use them pre-builded. Actually their update (once stabilized for sure) required only in case of moving into another release of Alpine.

True, ok, that's fine.
> 
> > * ARM64 hosts must be considered (so it will be needed to build x86 cross compiler), if you can't test at your side, let's sync up, I can find you people with Macs M1's and M2's to test your changes...
> 
> I finished experiments on Arm64 platform, resolved the bug and now it should work fine. But if you will find anybody with M1/M2, it would be great to check. Something like `make pkg/alpine pkg/cross-compilers&&make ZARCH=amd64 pkg/alpine pkg/kernel pkg/new-kernel pkg/pillar` should use cross-compilation

Great, I've asked someone to check, I'll bring the outcomes here...

> 
> > * Please, add information to your commit messages, your commits are: 1) introducing new packages; 2) introducing cross-compilers; 3) changing the way kernel is built.... so please, give as much information as possible regarding your changes so one doesn't need to go through the sources to understand the overview of the changes. Also, information about cross-compilation must be added to the documentation (I would recommend to add them to https://github.com/lf-edge/eve/blob/master/docs/BUILD.md)
> 
> Sure, will do.

rene commented on pkg/cross-compilers/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+FROM lfedge/eve-alpine:1fc941603ddc153f26104865db242e644c16aa99 as build-amd64
```
Don't we need --platform=${BUILDPLATFORM} here? It's building cross compiler that should run on the host, I'm wondering if some do a make ZARCH=arm64 ... pkg/cross-compilers on a x86 host... it will use emulation in this case...

giggsoff replied:
You are right, we need cross-compilers for the current platform we are sitting on (BUILDPLATFORM). But at the same time we want to have images pushed for all platforms in our CI. So we are forced to build them as we can (including emulation as we do not have riscv64 runner available). That is, when we run `make ZARCH=riscv64 pkgs` in CI we expect that it will construct riscv64-based image for us. The idea here is to not rebuild cross-compilers often.

rene replied:
@giggsoff , ok, got it.

rene commented:
@giggsoff , I'm still building the latest version of this PR on my side, but so far I don't have any other comments...

deitch commented:
At the request of @rene , I have been running a sample build, essentially:

```
make ZARCH=amd64 pkg/alpine
make pkg/cross-compilers
make ZARCH=amd64 pkg/kernel
```

I will post logs when they are done. In the meantime, I had a few questions.

First, while `make pkg/cross-compilers` does build only for my local platform, if I do pass it `ZARCH,` it gets messed up:

```sh
$ make -n pkg/cross-compilers ZARCH=linux/amd64
/home/deitch/go/src/github.com/lf-edge/eve/build-tools/bin/linuxkit  pkg build     --platforms linux/linux/amd64,linux/arm64 -build-yml build.yml pkg/cross-compilers
```

I am pretty sure that `--platforms` line is an error.

Second, what is the process for building across all platforms? If `pkg/cross-compilers` is an image, then it needs to be available for all build platforms, does it not?

deitch commented:
Build logs
[log.txt](https://github.com/lf-edge/eve/files/10319641/log.txt)

giggsoff commented:
> First, while `make pkg/cross-compilers` does build only for my local platform, if I do pass it `ZARCH,` it gets messed up:
> 
> ```shell
> $ make -n pkg/cross-compilers ZARCH=linux/amd64
> /home/deitch/go/src/github.com/lf-edge/eve/build-tools/bin/linuxkit  pkg build     --platforms linux/linux/amd64,linux/arm64 -build-yml build.yml pkg/cross-compilers
> ```
> 
> I am pretty sure that `--platforms` line is an error.

We want two have both platforms available: 

- the host one for using in build process of kernel
- the target one (provided by ZARCH) to be published (e.g. from CI)

We will remove it from `make pkgs` target (the same we did it for alpine-base) once published.

> Second, what is the process for building across all platforms? If `pkg/cross-compilers` is an image, then it needs to be available for all build platforms, does it not?

Probably I cannot understand the question, but yes, we want to push it for all platforms.

deitch commented:
> We want two have both platforms available:

Should one of them be `linux/linux/amd64`? 😁

giggsoff commented:
> > We want two have both platforms available:
> 
> Should one of them be `linux/linux/amd64`? grin

Ah, I see, odd...

giggsoff commented:
> > We want two have both platforms available:
> 
> Should one of them be `linux/linux/amd64`? grin

Wait, you defined `ZARCH=linux/amd64`, but it is `ZPLATFORM`...

deitch commented:
> Wait, you defined ZARCH=linux/amd64, but it is ZPLATFORM (I mean it should be ZARCH=amd64)

🤦‍♂️

giggsoff commented:
I put doccumentation and resolved all related yetus complains

deitch commented:
yetus still has 3 complaints: 2 about consecutive `RUN` statements, and 1 about using `WORKDIR`. Your judgment is superior to yetus, so if you think this is the right way, maybe put in the appropriate nolint instructions?

deitch commented on pkg/alpine/Dockerfile:
```diff
@@ -44,8 +44,20 @@ COPY eve-alpine-deploy.sh go-compile.sh /bin/
 
 RUN apk update && apk upgrade -a
 
+# define arch-specific envs
```
This is a nice, elegant solution.

deitch reviewed (changes requested):
Mostly on the documentation, some extra steps so that the next person can understand it.

I do think it is quite complex, so it will be hard to add these to more packages. There may be ways to simplify it, but that is for the future. If this works, all the better.

giggsoff commented on pkg/cross-compilers/build.yml:
```diff
@@ -0,0 +1,3 @@
+image: eve-cross-compilers
+org: lfedge
+network: yes
```
Not only. Aport scripts also use network to download sources of pacakges.

eriknordmark replied:
@giggsoff from a build reproducibility and SBOM perspective we want to avoid network: yes and instead rely on ADD commands. Is that possible for aports?

giggsoff replied:
I put required source archives into Dockerfile, let's see how it will pass CI.

giggsoff commented on docs/BUILD.md:
````diff
@@ -674,3 +674,110 @@ The latter aspect makes maintaining `lfedge/eve-alpine` a bit tricky, even thoug
+    RUN eve-alpine-deploy.sh
+    ```
+
+4. Adjust `TARGET_ARCH` environment variable for cross-compiler:
````
We have defaults and override explicitly only in case. Assume it is reasonable

giggsoff commented on docs/BUILD.md:
```diff
@@ -674,3 +674,110 @@ The latter aspect makes maintaining `lfedge/eve-alpine` a bit tricky, even thoug
+
+Cross-compilation support depends on compilers we use and ways of configuring compilers. The implementation of the process is quite varied and may not be possible. We should support both cases: native and cross-compilation for all targets.
+The main reason of enabling cross-compilation is to speedup build for another target architecture and acceleration can reach orders of magnitude.
+It is not reasonable to enable cross-compilation in all packages as it complicate the logic, only for ones which consume significant time.
```
Added, thank you.

giggsoff commented on pkg/cross-compilers/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+ENV APORTS /home/builder/aports
+
+# setting up building account
+RUN adduser -G abuild -D builder || : && \
```
Let's see how it works

giggsoff commented on docs/BUILD.md:
````diff
@@ -674,3 +674,110 @@ The latter aspect makes maintaining `lfedge/eve-alpine` a bit tricky, even thoug
+    RUN BUILD_PKGS="${BUILD_PKGS_BASE}" eve-alpine-deploy.sh
+    ```
+
+2. Use cross-compilers image for the host platform to install packages form it later:
````
correct, sorry

giggsoff commented on docs/BUILD.md:
```diff
@@ -674,3 +674,110 @@ The latter aspect makes maintaining `lfedge/eve-alpine` a bit tricky, even thoug
+with the same revision as used for eve-alpine package. It helps us to use the same versions of tools with the same configuration and patches applied.
+Only two patches applied to aports repository:
+
+1. [0001-only-cross-compile-prepare.patch](.../pkg/cross-compilers/patches/aports/0001-only-cross-compile-prepare.patch) - allow us to use the bootstrap script provided in repo and stops the process of bootstrap right after preparing of cross-compilers
```
I put comment about upstreaming.

giggsoff commented on docs/BUILD.md:
```diff
@@ -674,3 +674,110 @@ The latter aspect makes maintaining `lfedge/eve-alpine` a bit tricky, even thoug
+1. [0001-only-cross-compile-prepare.patch](.../pkg/cross-compilers/patches/aports/0001-only-cross-compile-prepare.patch) - allow us to use the bootstrap script provided in repo and stops the process of bootstrap right after preparing of cross-compilers
+2. [0002-adjust-sysroot.patch](.../pkg/cross-compilers/patches/aports/0002-adjust-sysroot.patch) - modifies sysroot configured during building of gcc to be pointed onto reasonable directory (e.g. `/usr/riscv64-alpine-linux-musl` instead of `/home/builder/sysroot-riscv64` bootstrap process use internally)
+
+The build process of the image is quite time-consuming, but we need to rebuild it only in case of moving onto the next alpine release (when we will update gcc/binutils). So it may be excluded from the CI for a while.
```
Added few details

giggsoff commented on docs/BUILD.md:
```diff
@@ -674,3 +674,110 @@ The latter aspect makes maintaining `lfedge/eve-alpine` a bit tricky, even thoug
+### cross-compilers package
+
+In order to enable cross-compilation we add `pkg/cross-compilers` which contains `binutils`, `gcc` and `g++` prepared with target architecture different from host one.
+We support `aarch64` and `riscv64` target architectures for `x86_64` host architecture and `x86_64` and `riscv64` target architectures for `aarch64` host architecture.
```
I put the table here.

> Also, is this just for kernel, or other packages as well? And for other packages, do we need the cross-compilers package if we are doing go? Or is it just if we use CGO?

It is about the whole set of packages, I did not tell about specific package here. In case of CGO, we need gcc with cross-compile support, and we should see into the table to understand, do we have support for our build-target pair.

eriknordmark reviewed (approved):
LGTM

rene commented:
@giggsoff , thanks for the updates, I don't have any other comments, LGTM.

giggsoff commented:
_rebased to resolve conflicts and update eve-alpine tag_

eriknordmark reviewed (approved):
Re-run eden

rene commented:
> Re-run eden

@eriknordmark , I've opened this PR as a new one: https://github.com/lf-edge/eve/pull/2998 

it's rebased against master branch again, please, re-run Eden to see if those errors were gone.

Let's discuss in the new one, we can close it this one.

giggsoff commented:
Replaced by #2998
